instana/instrumentation/google/cloud/pubsub.py

```diff
@@ -0,0 +1,106 @@
+        span.set_tag('gcps.op', 'consume')
+        # Fully qualified identifier is in the form of
+        # `projects/{project_id}/subscriptions/{subscription_name}`
+        project_id, subscription_id = subscription_path.split('/')[1::2]
+        span.set_tag('gcps.projid', project_id)
+        span.set_tag('gcps.sub', subscription_id)
+
+
+    @wrapt.patch_function_wrapper('google.cloud.pubsub_v1', 'PublisherClient.publish')
+    def publish_with_instana(wrapped, instance, args, kwargs):
+        """References:
+        - PublisherClient.publish(topic_path, messages, metadata)
+        """
+        # check if active
+        parent_span = tracer.active_span
+
+        # return early if we're not tracing
+        if parent_span is None:
+            return wrapped(*args, **kwargs)
+
+        with tracer.start_active_span('gcps-producer', child_of=parent_span) as scope:
+            # trace continuity, inject to the span context
+            trace_continuity_headers = dict()
+            tracer.inject(scope.span.context, Format.TEXT_MAP, trace_continuity_headers)
+
+            # update the metadata dict with instana trace attributes
+            kwargs['trace_continuity_headers'] = json.dumps(trace_continuity_headers)
+
+            _set_publisher_tags(scope.span, topic_path=args[0])
+
+            try:
+                rv = wrapped(*args, **kwargs)
+            except Exception as e:
+                scope.span.log_exception(e)
+                raise
+            else:
+                return rv
+
+
+    @wrapt.patch_function_wrapper('google.cloud.pubsub_v1', 'SubscriberClient.subscribe')
+    def subscribe_with_instana(wrapped, instance, args, kwargs):
+
+        """References:
+        - SubscriberClient.subscribe(subscription_path, callback)
+        - callback(message) is called from the subscription future
+        """
+
+        def callback_with_instana(message):
+            attr = message.attributes
+            if attr:
+                # trace continuity
+                headers = json.loads(attr.get('trace_continuity_headers'))
+                parent_span = tracer.extract(Format.TEXT_MAP, headers)
+            else:
+                parent_span = None
+
+            with tracer.start_active_span('gcps-consumer', child_of=parent_span) as scope:
+                _set_consumer_tags(scope.span, subscription_path=args[0])
+                try:
+                    callback(message)
+                except Exception as e:
+                    scope.span.log_exception(e)
+                    raise
+
+        # Handle callback appropriately from args or kwargs
+        if 'callback' in kwargs:
+            callback = kwargs['callback']
+            kwargs['callback'] = callback_with_instana
+            return wrapped(*args, **kwargs)
+        else:
+            callback = args[1]
+            args_list = list(args)
+            args_list[1] = callback_with_instana
+            args = tuple(args_list)
```

Comment:
Would it work to do something like:
```python
subscription, callback, *args = args
```
and than
```python
args = (subscription, callback_with_instana, *args)
```
? This way you could use `subscription` and `callback` instead of `args[0]` in the callback wrapper.

Reply:
Thanks, this will work - much neater ;)